common/src/test/java/io/netty/util/internal/logging/InternalLoggerFactoryTest.java

```diff
@@ -230,4 +246,142 @@ public void shouldNotAllowToSetFactoryTwice() {
+        final CyclicBarrier barrier = new CyclicBarrier(3);
+        final InternalLoggerFactory.InternalLoggerFactoryHolder holder =
+                new InternalLoggerFactory.InternalLoggerFactoryHolder();
+        final InternalLoggerFactory first = createMock(InternalLoggerFactory.class);
+        final InternalLoggerFactory second = createMock(InternalLoggerFactory.class);
+
+        final AtomicBoolean firstSetSuccess = new AtomicBoolean();
+        final Thread firstSet = setThread(first, holder, firstSetSuccess, barrier);
+
+        final AtomicBoolean secondSetSuccess = new AtomicBoolean();
+        final Thread secondSet = setThread(second, holder, secondSetSuccess, barrier);
+
+        firstSet.start();
+        secondSet.start();
+        // start the two set threads
+        barrier.await();
+
+        // wait for the two set threads to complete
+        barrier.await();
+
+        assertTrue(firstSetSuccess.get() || secondSetSuccess.get());
+        if (firstSetSuccess.get()) {
+            assertFalse(secondSetSuccess.get());
+            assertSame(first, holder.getFactory());
+        } else {
+            assertFalse(firstSetSuccess.get());
+            assertSame(second, holder.getFactory());
+        }
+    }
+
+    private static Thread getThread(
+            final AtomicReference<InternalLoggerFactory> reference,
+            final InternalLoggerFactory.InternalLoggerFactoryHolder holder,
+            final CyclicBarrier barrier) {
+        return new Thread(new Runnable() {
+            @Override
+            public void run() {
+                awaitUnchecked(barrier);
+                reference.set(holder.getFactory());
+                awaitUnchecked(barrier);
+            }
+        });
+    }
+
+    private static Thread setThread(
+            final InternalLoggerFactory internalLoggerFactory,
+            final InternalLoggerFactory.InternalLoggerFactoryHolder holder,
+            final AtomicBoolean setSuccess,
+            final CyclicBarrier barrier) {
+        return new Thread(new Runnable() {
+            @Override
+            public void run() {
+                awaitUnchecked(barrier);
+                boolean success = true;
+                try {
+                    holder.setFactory(internalLoggerFactory);
+                } catch (final IllegalStateException e) {
+                    success = false;
+                    assertThat(e.getMessage(), containsString("factory is already set"));
+                } finally {
+                    setSuccess.set(success);
+                    awaitUnchecked(barrier);
+                }
+            }
+        });
+    }
+
+    private static void awaitUnchecked(CyclicBarrier barrier) {
+        try {
+            barrier.await();
+        } catch (InterruptedException exception) {
+            throw new RuntimeException(exception);
+        } catch (BrokenBarrierException exception) {
+            throw new RuntimeException(exception);
```

Comment:
Okay. I'll address this when I force push a squash commit before integration, if that's okay with you @normanmaurer?